### PR TITLE
New Drawer component

### PR DESCRIPTION
**Pull request by pkwi**
A drawer is a UI component that overlays on top of an existing UI to display additional content.

### CODE REVIEW

File: packages/core/src/components/drawer/drawer.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        }
+    }
+
+    private maybeRenderCloseButton() {
+        // show close button if prop is undefined or null
+        // this gives us a behavior as if the default value were `true`
+        if (this.props.isCloseButtonShown !== false) {
+            return (
+                <button
```
Comment (giladgray) [inline packages/core/src/components/drawer/drawer.tsx]: use `<Button minimal>` like we did in Dialog

File: packages/docs-app/src/examples/core-examples/drawerExample.tsx [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                                transformation authoring, and more. But these powers are not limited to the back-end IT
+                                shop.
+                            </p>
+                            <p>
+                                In Foundry, tables, applications, reports, presentations, and spreadsheets operate as
+                                data integrations in their own right. Access controls, transformation logic, and data
+                                quality flow from original data source to intermediate analysis to presentation in real
+                                time. Every end product created in Foundry becomes a new data source that other users
+                                can build upon. And the enterprise data foundation goes where the business drives it.
+                            </p>
```
Comment (giladgray) [inline packages/docs-app/src/examples/core-examples/drawerExample.tsx]: this is far too much content! maybe set a max-width of like 700 on a centered container inside the body?

File: packages/core/src/components/drawer/_drawer.scss [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      "exit",
+      (transform: (translateY(100%), translateY(0))),
+      $pt-transition-duration,
+      $before: "&"
+    );
+    top: auto;
+    bottom: 0;
+    left: 0;
+    width: 100vw;
+    height: 50vh;
```
Comment (giladgray) [inline packages/core/src/components/drawer/_drawer.scss]: i was wrong about viewport units, this actually works much better with %s. disable the "Use `Portal`" switch and you'll see what i mean. it's way too big with vh/w but works perfectly when changed to %.

File: packages/core/src/components/drawer/_drawer.scss [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  );
+  display: flex;
+  flex-direction: column;
+  top: 0;
+  right: 0;
+  bottom: 0;
+  margin: 0;
+  box-shadow: $pt-elevation-shadow-4;
+  background: $white;
+  width: 50vw;
```
Comment (giladgray) [inline packages/core/src/components/drawer/_drawer.scss]: `50%`

File: packages/docs-app/src/examples/core-examples/drawerExample.tsx (hunk 2 above)
Comment (pkwi) [reply packages/docs-app/src/examples/core-examples/drawerExample.tsx]: ah my fault. I doubled the content to make it scrollable

File: packages/core/src/components/drawer/drawer.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+@# Drawer
```
Comment (giladgray) [inline packages/core/src/components/drawer/drawer.md]: mark it as `new`
```suggestion
---
tag: new
---

@# Drawer
```

File: packages/core/src/components/drawer/drawer.tsx [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+     * Toggles the visibility of the overlay and its children.
+     * This prop is required because the component is controlled.
+     */
+    isOpen: boolean;
+
+    /**
+     * Drawer always has a backdrop so this prop is excluded from the public API.
+     * @internal
+     */
+    hasBackdrop?: boolean;
```
Comment (giladgray) [inline packages/core/src/components/drawer/drawer.tsx]: remove this prop, or edit the docs to note default based on size.

File: packages/core/src/components/drawer/drawer.tsx [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+     * Whether the drawer should appear with vertical styling.
+     * @default false
+     */
+    vertical?: boolean;
+
+    /** Whether this drawer should use large styles. */
+    large?: boolean;
+
+    /** Whether this drawer should use small styles. */
+    small?: boolean;
```
Comment (giladgray) [inline packages/core/src/components/drawer/drawer.tsx]: alphabetize these two new props

File: packages/core/src/components/drawer/drawer.tsx [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    /**
+     * Whether the drawer should appear with vertical styling.
+     * @default false
+     */
+    vertical?: boolean;
+
+    /** Whether this drawer should use large styles. */
+    large?: boolean;
+
+    /** Whether this drawer should use small styles. */
```
Comment (giladgray) [inline packages/core/src/components/drawer/drawer.tsx]: more detail as to what small/large mean here - 25% / 75%

File: packages/core/src/components/drawer/drawer.tsx [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+                <Overlay {...this.props} className={Classes.OVERLAY_CONTAINER} hasBackdrop={false}>
+                    <div className={classes} style={this.props.style}>
+                        {this.maybeRenderHeader()}
+                        {this.props.children}
+                    </div>
+                </Overlay>
+            );
+        }
+        return (
+            <Overlay {...this.props} className={Classes.OVERLAY_CONTAINER} hasBackdrop={true}>
```
Comment (giladgray) [inline packages/core/src/components/drawer/drawer.tsx]: allow user to override, with default behavior. 
```suggestion
            <Overlay {...this.props} className={Classes.OVERLAY_CONTAINER} hasBackdrop={this.props.hasBackdrop || !this.props.small}>
```
then remove `if` block above.

**Comment by giladgray** [comment]
needs tests and polish.

**Comment by ptvandi** [comment]
I just started to build this myself and am really excited to have found that it's in the works. Thank you and well done!

**Comment by mltejera** [comment]
@giladgray Just got my hands on this! The timing is great as my team just started rolling our own drawers earlier this week.

Do you know if there are plans to make it open from the top or left? Thanks!

**Comment by Darkness126** [comment]
Is it possible for the drawer to be on the left side? 
I was able to get the drawer to left by setting style left:0 but the animation of closing and opening is happening towards the right.

Please help

**Comment by adidahiya** [comment]
@Darkness126 check out the open issues, there are a few open ones about that #3360. It's almost always better to file a new issue rather than comment on a closed PR.

**Pull request merged**